### PR TITLE
Add extension point for the non-GET dialog

### LINKED ISSUE #63: Add extension points to render custom POST form
@mikekelly Spring Data REST is using this project (via webjars) as a plugin, so anybody that builds a HAL-based service can easily serve up a HAL Browser front end with next to no effort.

A bonus feature Spring Data REST comes with is ALPS metadata which includes the properties for every resource. A feature I have wanted to add for some time is to use this metadata and make the HAL Browser's POST form more sophisticated by listing each property by name with separate fields, instead of a raw body field where the user must type in all the fields in JSON, and include the right names.

I am tinkering with a fork of the app to craft some extension point where someone could drop in extra JavaScript to fetch this extra metadata, but I'm finding it tricky to do so. I didn't know if you were interested in working on such an extension point for this project, or if you are too busy with other projects.

/cc @olivergierke

### PR DESCRIPTION
**Pull request by gregturn**
This patch provides the means to inject a different non-GET dialog. The default dialog requires useres to type in a JSON document for POST support, requiring the user to know what is required. By offering an extension, developers can code a more intelligent pop-up based upon alternative metadata not supplied by this module.

After this is merged, and I can use this in my own project, I am happy to update the documentation with a polished example of using this extension.

Resolves #63

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,15 +1,15 @@
-HAL-browser
-===========
+# HAL-browser
```
Comment (mikekelly) [inline README.md]: :+1: !

File: README.md [hunk 2]
```diff
@@ -25,16 +25,87 @@ representing links. For example:
...
+
+If you have some form of metadata (such as an ALPS micro-document), you can plug in your own logic and render a more user-friendly dialog with each attribute listed.
+
+The following fragment of JavaScript shows the API to plugin your own solution:
+```javascript
+HAL.customPost = {
+
+	renderForm: function(href, callback) {
+		// 1. var results = /* array of attribute names ["firstName", "lastName"] */
+		// 2. invoke callback(results)
```
Comment (mikekelly) [inline README.md]: It feels like this could be more general ie. the callback could take a string of html

File: README.md [hunk 3]
```diff
@@ -25,16 +25,87 @@ representing links. For example:
...
+	},
+
+	extractData: function(attributes) {
```
Comment (mikekelly) [inline README.md]: not a blocker, but it might be more intention revealing to call this `handleFormSubmission` or similar

Comment (mikekelly) [inline README.md]: or `onFormSubmit`

**Comment by joshco** [comment]
This is a really good idea.  
Can you include a complete example for reference?
Does it handle nested attributes?  Suppose my post needs to be:
`````
"person" : {
   "first_name" : "Josh",
   "last_name" : "Cohen"
}
````

**Comment by mikekelly** [comment]
thinking about this, another approach is just to let people specify a `customPostForm` which is a `Backbone.View` that takes an options object that includes the properties of the link (ie. the link relation url and the target url of the link as well)

```javascript
CustomPostForm = Backbone.View.extend({
initialize: function(opts) {
    this.link_rel_href = opts.link_rel_href;
  },
});

HAL.customPostForm = CustomPostForm;
```

**Comment by gregturn** [comment]
Here is my example, (which hasn't been updated to match the new function names yet in this PR) => https://github.com/gregturn/hal-browser/blob/alps/js/custom.js#L5-L48

I have an ALPS document available from the same place as my RESTful service, so most of the code involves doing another ajax call, and then using JSONPath to create an array of values => ["firstName", "lastName"].

Then, I use it in a different template => https://github.com/gregturn/hal-browser/blob/alps/browser.html#L183-L196

This variant of the template swaps out the Body textarea, and instead iterates over ["firstName", "lastName"] to generate an array of <input> fields.

Finally, my extraction code just uses jQuery to pull out the values, assembled a JSON object, and stringify them.

File: README.md (hunk 2 above)
Comment (gregturn) [inline README.md]: Perhaps. I have pasted below the flow of the parameter fed to callback().

Flow of the callback:
1. https://github.com/gregturn/hal-browser/blob/add-extension-point/js/hal/views/links.js#L41-L46
2. https://github.com/gregturn/hal-browser/blob/add-extension-point/js/hal/views/non_safe_request_dialog.js#L5
3. https://github.com/gregturn/hal-browser/blob/add-extension-point/js/hal/views/non_safe_request_dialog.js#L53
4. https://github.com/gregturn/hal-browser/blob/add-extension-point/js/hal/views/non_safe_request_dialog.js#L28-L30

**Comment by gregturn** [comment]
@mikekelly An important thing I want to get in here is having not only the href of the link, but also the name of the rel. There is much metadata invested in rel names for Spring Data REST.

**Comment by mikekelly** [comment]
> I have an ALPS document available from the same place as my RESTful service, so most of the code involves doing another ajax call, and then using JSONPath to create an array of values => ["firstName", "lastName"].

This is my concern. I think an array-of-fields interface here is not general enough. The relation URL may have a complex schema document specifying types of input, value ranges, enums, etc.

This is what led to my above proposal; it seems like this could me modelled as a custom view that can be injected.

**Comment by gregturn** [comment]
See #66 for my updated version. @mikekelly based on your suggestion of creating an alternative Backbone view, I found that approach much simpler and less invasive. Hopefully we can push this through quicker.

**Pull request closed without merging**